local isomorphisms of graphs

avekens opened the pull request:
Tis is the fourth item of issue #4808.

* ~ el3v3  moved from mathbox to main
* ~fabexd new in main (proof of ~fabexg  and ~f1oabexg  shortened)
* New theorems about isomorphisms of graphs in AV's mathbox: ~gricrcl, ~isubgrgrim, ~uhgrimisgrgric, ~clnbgrisubgrgrim, ~clnbgrgrim
* Definition and basic theorems for local isomorphisms of graphs in AV's mathbox

wlammen commented on set.mm:
```diff
@@ -78255,11 +78289,19 @@ Axiom of Replacement (unlike ~ resfunexg ).  (Contributed by NM,
 
   ${
-    $d A f $.  $d B f $.
+    $d A f $.  $d B f $.  $d C f $.  $d D f $.
```
The OLD proof depends on fewer dv conditions.  I understand that f is bounded, and that a change of variable like 
{ f | ( f : A -1-1-onto-> B /\ ph ) } = { g | ( g : A -1-1-onto-> B /\ [ g / f ] ph ) }
holds, but that is hardly provable without ax-12.

Note that the old proof does not depend on ax-10 .. ax-12.
I think this is not a valid shortening.

avekens replied:
The new proofs neither depend on ax-10, ax-11, ax-12. 
For all of the four theorems fabexg, fabexgOLD, f1oabexg, f1oabexgOLD, calling `sh tr xxx /ax/match ax-*`  results in 
`ax-mp ax-1 ax-2 ax-3 ax-gen ax-4 ax-5 ax-6 ax-7 ax-8 ax-9 ax-ext ax-sep ax-nul ax-pow ax-pr ax-un`

wlammen replied:
It is not the axiom list I am not satisfied with.

The expression A e. C is a convenient shorthand for E. x x = A, or by isset for A e. V_.  So a fix would simply use A e. V_ and B e. V_ in the proof first, and then in a final step relax these expressions to A e. C and B e. D by elisset.

C and D are not subject to any restriction to fulfill their role. Here you artificially add some.  Get rid of them.

wlammen commented on set.mm:
```diff
@@ -810491,6 +810542,678 @@ as an isomorphism in the category of graphs (something like "GraphIsom
+  neighborhoods.
+
+  Additionally, a binary relation ` ~=lgr ` is defined (see ~ df-grlic ) which
+  is true for two graphs iff there is a local isomorphisms between these
```
Non-blocking
typo:
there are local isomorphisms
or
there is a local isomorphism

tirix commented on set.mm:
```diff
@@ -810491,6 +810542,678 @@ as an isomorphism in the category of graphs (something like "GraphIsom
+
+  ${
+    $d f g h v $.
+    $( A local isomorphism of graphs is a bijection between the sets of
+       vertices of two graphs that preserves local adjacency, i.e. that the
+       subgraph induced by the closed neighborhood of a vertex of the first
+       graph and the subgraph induced by the closed neighborhood of the
+       associated vertex of the second graph are isomorphic.  See the following
+       chat in mathoverflow:
```
I.e. _such_ that the subgraph... ?

tirix commented on set.mm:
```diff
@@ -810491,6 +810542,678 @@ as an isomorphism in the category of graphs (something like "GraphIsom
+  $}
+
+
+  ${
+    $d V x y z $.  $d .~ x y z $.
+    brinxper.r $e |- ( x e. V -> x .~ x ) $.
+    brinxper.s $e |- ( x e. V -> ( x .~ y -> y .~ x ) ) $.
+    brinxper.t $e |- ( x e. V -> ( ( x .~ y /\ y .~ z ) -> x .~ z ) ) $.
+    $( Conditions for a reflexive, symmetric and transitive binary relation to
+       be an equivalence relation over a class ` V ` .  (Contributed by AV,
+       11-Jun-2025.) $)
+    brinxper $p |- ( .~ i^i ( V X. V ) ) Er V $=
+      ( cv wbr wa wcel wi brxp adantr sylbi impcom brin com12 adantl cxp sylbb2
+      cin relinxp ancom jctird 3imtr4i expd simplr simprl anim12ci exp31 anbi2i
+      imp jca 3imtr4g sylibr id sylanbrc simplbi impbii bitr4i iseri ) ABCEDEEU
+      AZUCZEEDUDAIZBIZDJZVFVGVDJZKZVGVFDJZVGVFVDJZKZVFVGVEJZVGVFVEJVIVHVMVIVFEL
+      ZVGELZKZVHVMMVFVGEENZVQVHVKVLVOVHVKMVPGOVQVPVOKVLVOVPUEVGVFEENUBUFPQVFVGD
+      VDRZVGVFDVDRUGVNVGCIZVEJZKVFVTDJZVFVTVDJZKZVFVTVEJVNWAWDVNVJWAWDMVSWAVJWD
+      WAVGVTDJZVGVTVDJZKZVJWDMVGVTDVDRWGVHVQKZWBVOVTELZKZKZVJWDWFWEWHWKMZWFVPWI
+      KZWEWLMVGVTEENWMWEWHWKWMWEKZWHKWBWJWNWHWBWEWHWBMWMWHWEWBVQVHWEWBMZVOVHWOM
+      VPVOVHWEWBHUHOQSTUNWNWIWHVOVPWIWEUIVHVOVPUJUKUOULPQVIVQVHVRUMWCWJWBVFVTEE
+      NUMUPPSPUNVFVTDVDRUQVOVFVFDJZVFVFVDJZKZVFVFVEJVOWRVOWPWQFVOVOVOWQVOURZWSV
+      FVFEENZUSUOWQVOWPWQVOVOWTUTTVAVFVFDVDRVBVC $.
+  $}
```
This is a generic theorem, it could be moved to a section specific to relations.

avekens commented:
With commit [3aa1d22], I modified the proofs for ~fabexg and ~f1oabexg so that they do not require addititional DV conditions, and were a little bit longer as before, but still shorter then the original proofs.
Then, with commit [2e068b6], I moved up ~mapex, as suggested by @icecream17 , but used ~fabexg to shorten its proof. Now all these similar theorems are at a single place, are based on the same axioms, and have no additional DV conditions.

It seems that ~mapex is the theorem which should be called "fabexg", because it is about a class abstraction containing functions. The other theorems are derivatives of it, and may be called "fimabexd" (instead of ~fabexd, meaning "class abstraction implied by a function") and "fanabexg" (instead of ~fabexg, meaning "class abstraction with a function and another restriction"). But this is a different story...
